HLTrigger/HLTcore/test/test_catch2_TriggerExpressionParser.cc

```diff
@@ -0,0 +1,52 @@
+#define CATCH_CONFIG_MAIN
+#include "catch.hpp"
+
+#include <string>
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "HLTrigger/HLTcore/interface/TriggerExpressionParser.h"
+
+namespace {
+
+  bool testExpression(std::string const& expression) {
+    auto const* expr = triggerExpression::parse(expression);
+
+    if (not expr) {
+      edm::LogWarning("InvalidInput") << "Couldn't parse trigger results expression \"" << expression << "\"";
+      return false;
+    }
+
+    edm::LogPrint("testExpression") << "Parsed expression: \"" << expression << "\"";
+    return true;
+  }
+
+}  // namespace
+
+TEST_CASE("Test TriggerExpressionParser", "[TriggerExpressionParser]") {
+  // examples of expressions supported by the triggerExpression parser
+  SECTION("CorrectExpressions") {
+    REQUIRE(testExpression("TRUE"));
+    REQUIRE(testExpression("FALSE"));
+    REQUIRE(testExpression("NOT (FALSE)"));
+    REQUIRE(testExpression("(NOT FALSE) OR TRUE"));
+    REQUIRE(testExpression("NOTThisHLTPath AND TRUE AND NOT L1_A?_*"));
+    REQUIRE(testExpression("NOT NOTThisHLTPath"));
+    REQUIRE(testExpression("ThisHLTANDNOTThatORTheOther"));
+    REQUIRE(testExpression("NOT L1_SEED1 AND L1_SEED2*"));
+    REQUIRE(testExpression("NOT L1_SEED2 AND (HLT_PATH_? AND NOT HLT_PATH2_??_*)"));
+    REQUIRE(testExpression("NOT (HLT_Path1 AND HLT_Path2)"));
+    REQUIRE(testExpression("NOT (NOTHLT_Path OR HLT_Path2)"));
+    REQUIRE(testExpression("((L1_A AND HLT_B) OR Dataset_C) AND NOT (Status_D OR Name_E OR HLT_F*) AND L1_??_?_?"));
+    REQUIRE(testExpression("NOT (NOT (HLT_Path1 AND HLT_Path_*))"));
+  }
+
+  // examples of expressions not supported by the triggerExpression parser
+  SECTION("IncorrectExpressions") {
+    REQUIRE(not testExpression("A | B"));
+    REQUIRE(not testExpression("A && B"));
+    REQUIRE(not testExpression("NOT L1_SEED1 ANDD L1_SEED2*"));
+    REQUIRE(not testExpression("NOT (NOTHLT_Path OR HLT_Path2))"));
+    REQUIRE(not testExpression("NOT NOT (HLT_Path1 AND L1_Seed_?? OR HLT_Path_*)"));
+    REQUIRE(not testExpression("HLT_Path* NOT TRUE"));
+  }
```

Comment:
This list of good and bad expressions is pretty much random. Suggestions are welcome.

Reply:
Why is this one invalid ?
```
testExpression("NOT NOT (HLT_Path1 AND L1_Seed_?? OR HLT_Path_*)"));
```

I mean, it _is_ silly to use `NOT NOT`, but is it an error?

Reply:
Ideally, the parser should understand it, but it doesn't.. `NOT NOT [expr]` isn't working, while `NOT (NOT [expr])` will work (last valid expression in the unit test). In the first case, it again gets into a state where it parses one of the `NOT`s (I assume the 2nd one) as an HLT-path name, then fails.

Reply:
Understood.

OK, I think `NOT NOT ...` is pretty low on the wish list, we can definitely live without.

Reply:
Okay, I would agree. On the other hand, more checking led me to realise that the parser reads `NOTHLT1` as equivalent to `NOT HLT1` (or `NOT(HLT1)`, or `NOT (HLT1)`), and that seems wrong.. so I'll have to take a deeper look.

Reply:
Maybe `HLT1 ANDHLT2` _etc_ share the same problem ?

Reply:
Indeed, they do.. the best I could come up with right now is the following:
```diff
-      unary = ((qi::lit("NOT") >> operand)[qi::_val = new_<OperatorNot>(qi::_1)] | operand[qi::_val = qi::_1]);
+      unary = ((qi::lit("NOT ") >> operand)[qi::_val = new_<OperatorNot>(qi::_1)] | operand[qi::_val = qi::_1]);

      expression =
-          unary[qi::_val = qi::_1] >> *((qi::lit("AND") >> unary)[qi::_val = new_<OperatorAnd>(qi::_val, qi::_1)] |
+          unary[qi::_val = qi::_1] >> *((qi::lit("AND ") >> unary)[qi::_val = new_<OperatorAnd>(qi::_val, qi::_1)] |
-                                        (qi::lit("OR") >> unary)[qi::_val = new_<OperatorOr>(qi::_val, qi::_1)]);
+                                        (qi::lit("OR ") >> unary)[qi::_val = new_<OperatorOr>(qi::_val, qi::_1)]);
```
That seems to work (more checks tomorrow), but it would make the following invalid b/c of no spaces
```
(ABC)AND((THIS)OR(THAT))
```